### PR TITLE
A4a: Streamline sign-up flow - Redirect logged out users to WPCOM

### PR DESCRIPTION
**Pull request by gogdzl**
Related to https://github.com/Automattic/automattic-for-agencies-dev/issues/352
Related to https://github.com/Automattic/automattic-for-agencies-dev/issues/353

## Proposed Changes

* This PR adds the logic necessary to redirect logged-out users to WPCOM based on their email:
   * Existing WPCOM users are redirected to the login page
   * Non-existing WPCOM users are redirected to WPCOM's signup page

## Testing Instructions

**NOTE:** This doesn't work for `automattic.com` emails, please use another email domain for testing.

* Apply the diff D146625-code to your sandbox
* Point `public-api.wordpress.com` in your hosts file to your sandbox
* Pull this branch and restart a8c-for-agencies ( `yarn start-a8c-for-agencies` )
* Make sure you are logged out of WPCOM (using a private browser window helps here)
* Go to the signup page (http://agencies.localhost:3000/signup)
* Fill in the form using an existing WPCOM email
  * Verify that you are redirected to the WPCOM login page
  * Log in and authorize A4A
  * Verify that after authorization you are redirected to `/signup/resume` and your form is displayed as a JSON string on the page
* Fill in the form using a non-existing WPCOM email
  * Verify that you are redirected to the WPCOM signup page
  * Create a new account and authorize A4A
  * Verify that after authorization you are redirected to `/signup/resume` and your form is displayed as a JSON string on the page

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx [hunk 1]
```diff
@@ -18,6 +20,51 @@ import type { APIError } from 'calypso/state/a8c-for-agencies/types';
 
 import './style.scss';
 
+async function handleWPCOMRedirect( payload: AgencyDetailsPayload, userLocale: string ) {
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: Maybe we could move this logic to a separate hook? What do you think?

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx [hunk 2]
```diff
@@ -18,6 +20,51 @@ import type { APIError } from 'calypso/state/a8c-for-agencies/types';
...
+	try {
+		const userValidationResponse = await wpcom.req.post( '/signups/validation/user', {
+			email: payload.email,
+		} );
+
+		const isNewUser = userValidationResponse && userValidationResponse.success;
+		const returnUri = new URL( '/signup/resume', window.location.origin );
+		const authUrl = new URL( config( 'wpcom_authorize_endpoint' ) );
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: I'm still not very strong in the context. Could we instead redirect to a4a `/connect` path, which handles auth?

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx [hunk 3]
```diff
@@ -18,6 +20,51 @@ import type { APIError } from 'calypso/state/a8c-for-agencies/types';
...
+			response_type: 'token',
+			client_id: config( 'oauth_client_id' ),
+			redirect_uri: returnUri.toString(),
+			scope: 'global',
+		} ).toString();
+
+		let wpcomRedirectUrl = undefined;
+
+		if ( isNewUser ) {
+			wpcomRedirectUrl = new URL(
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: It seems we could refactor this a bit, as the only thing that differs is URL and response type. What do you think?

File: client/a8c-for-agencies/sections/signup/controller.tsx [hunk 4]
```diff
@@ -1,9 +1,20 @@
...
+import { getSignupDataFromLocalStorage } from './lib/signup-data-to-local-storage';
 import AgencySignUp from './primary/agency-signup';
 
 export const signUpContext: Callback = ( context, next ) => {
 	context.store.dispatch( hideMasterbar() );
 	context.primary = <AgencySignUp />;
 	next();
 };
+
+export const signUpResumeContext: Callback = ( context, next ) => {
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/signup/controller.tsx:12]: I could handle that in my follow PR. You could either leave it as is, or just remove for now.

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx [hunk 5]
```diff
@@ -18,6 +20,51 @@ import type { APIError } from 'calypso/state/a8c-for-agencies/types';
...
+				redirect_to: authUrl.toString(),
+				email_address: payload.email ?? '',
+			} ).toString();
+		}
+
+		window.location.href = wpcomRedirectUrl.toString();
+	} catch ( e ) {
+		// TODO: Handle error.
+		// eslint-disable-next-line no-console
+		console.log( e );
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: I assume we could do something like `dispatch( errorNotice( error?.message, { id: notificationId } ) );` as earlier in the file.

**Review by andrii-lysenko — COMMENTED** [review]
Thank you for working on this. I left a few comments/questions.

Also, while testing, when creating a new account we need to undergo an email verification. From there, it's actually taking us to Wordpress.com. I wonder if there's a way we could have a proper redirect in the email to a4a after login. I assume thought it might be not a trivial task and couldn't be added to this project.

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx (hunk 3 above)
Comment (gogdzl) [reply client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: The params also differ.

File: client/a8c-for-agencies/sections/signup/controller.tsx (hunk 4 above)
Comment (gogdzl) [reply client/a8c-for-agencies/sections/signup/controller.tsx:12]: That's the idea; I left only a placeholder there for you to fill in later in a PR.

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx (hunk 2 above)
Comment (gogdzl) [reply client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: Nice idea! We can create this as a new issue to make this as a future improvement/refactor.

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx (hunk 5 above)
Comment (gogdzl) [reply client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: Addressed in 9335fbe

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx (hunk 1 above)
Comment (gogdzl) [reply client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: Addressed in 9335fbe

File: client/a8c-for-agencies/sections/signup/signup-form/index.tsx (hunk 2 above)
Comment (gogdzl) [reply client/a8c-for-agencies/sections/signup/signup-form/index.tsx]: Issue created here: https://github.com/Automattic/automattic-for-agencies-dev/issues/383

**Review by andrii-lysenko — APPROVED** [review]
Thanks for addressing comments. Look good to me!

**Pull request merged**